guides/hack/10-types/56-darray-varray-runtime-options.md

```diff
@@ -0,0 +1,69 @@
+
+The second situation, calling unset on an element of a `varray<_>`, can have multiple intends too.
+
+ - If the T is a nullable type, the programmer might have meant to write `null` to the index. This is more common in code written before hhvm4.
+ - The programmer does not care about the keys. The array is merely a meant to be used as a `KeyedContainer<not_important, T>` and he or she just meant to remove the value from the `KeyedContainer<_, _>`.
+ - The programmer intended to unset the last index.
+
+The first case is most likely a confusion caused by a removed behavior of all legacy arrays. Before hhvm 4 accessing an key that wasn't present would log a notice and return null. An unset on an array would under these circumstances act very similarly to explicitly setting to value to null. This is however quite tricky to do right if this array is being passed around the program a lot. An unset key is actually removed from the array. This means that `C\contains_key()` will return `false`, `idx()` will return its default argument, and `??` will evaluate to the RHS. However, explicitly setting the value to `null` does not remove the key from the array. This means that `C\contains_key()` returns `true`, `idx()` will return the `null`, but `??` will be unaffected.
+
+The second usecase is not met by Hack arrays. There is no `Container<_>` type that allows you to append to the end and remove things by index (except for `keyset<_>`, but that has a constraint value type). You can however emulate this behavior using a `varray<_>` or `vec<_>`. Removing the first key can be done using `Vec\drop($x, 1)`. Removing the last key can be done using `Vec\take($x, C\count($x) - 2)`. Removing a key from the middle can be done with the slightly unwieldly `Vec\filter_with_key($x, ($key, $_) ==> $key === 1)`. All of these will rekey the array. Any values after the key will be shifted down. This does however have a computational complexity of `O(n)`. If you need to remove a lot of keys from the middle that are next to eachother, use `Vec\slice()` to save some resources. If you need to remove a ton of arbitrary keys, at different points of your function it might be better to `dict($x)`, unset on the `dict<_, _>` and rekey it back to a `varray<_>` using `varray()` or `array_values()` depending on your hhvm version.
+
+The third usecase used to be valid Hack. Unsetting the last index of a `varray<_>` or `vec<_>` was allowed and acted like an <oss>\array_pop()</oss><facebook>C\fb\pop_back()</facebook> (THIS IS NOT ACUTALLY VALID SYNTAX FOR DISABLING TEXT IN OSS OR FB). This will currently not generate a warning, but it is unclear to me if this will continue to be allowed. The typechecker already raises a typeerror when you use unset on a non dictionary/hashmap like array type.
```

Comment:
I would live to use `\array_pop()` on hhvm.com, but have `C\fb\pop_back()` show in Facebook's internal documentation.

Reply:
```
`\array_pop()` <span class="fbonly apiAlias">`C\fb\pop_back`</span>
```